### PR TITLE
[ADD][11.0] account_global_discount: New Module

### PR DESCRIPTION
**Pull request by chienandalu**
Apply global discounts to invoices

Depends on:
- [x] base_global_discount: https://github.com/OCA/server-backend/pull/43

cc @Tecnativa

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Add a WIP commit with temp branch for being able to test this PR.

**Comment by chienandalu** [comment]
@pedrobaeza Done

**Comment by chienandalu** [comment]
travis is building fine (although failing in tests), but I can't get runbot to do so...

**Comment by pedrobaeza** [comment]
@moylop260 can you please help here that there's no chance of getting runbot working in this PR?
```
Unable to find image 'oca-account-invoicing:a04cd49_2' locally
docker: Error response from daemon: repository oca-account-invoicing not found: does not exist or no pull access.
See 'docker run --help'.
```

**Comment by moylop260** [comment]
FYI if the image is not found creating a container from `Full all logs` so the building of image (previous step) was wrong.

You can check a previous log `Full base logs` where `Docker build image` process is logged from 
 - Screen Shot 2019-08-23 at 10 38 20 AM


There are a problem of dependency 
```bash
Collecting mysqlclient (from -r /home/odoo/dependencies/server-backend/requirements.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/4d/38/c5f8bac9c50f3042c8f05615f84206f77f03db79781db841898fde1bb284/mysqlclient-1.4.4.tar.gz (86kB)
[91m    ERROR: Command errored out with exit status 1:
     command: /.repo_requirements/virtualenv/python3.5/bin/python3.5 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-5gkhsryw/mysqlclient/setup.py'"'"'; __file__='"'"'/tmp/pip-install-5gkhsryw/mysqlclient/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base pip-egg-info
         cwd: /tmp/pip-install-5gkhsryw/mysqlclient/
    Complete output (12 lines):
    /bin/sh: 1: mysql_config: not found
    /bin/sh: 1: mariadb_config: not found
    /bin/sh: 1: mysql_config: not found
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-5gkhsryw/mysqlclient/setup.py", line 16, in <module>
        metadata, options = get_config()
      File "/tmp/pip-install-5gkhsryw/mysqlclient/setup_posix.py", line 61, in get_config
        libs = mysql_config("libs")
      File "/tmp/pip-install-5gkhsryw/mysqlclient/setup_posix.py", line 29, in mysql_config
        raise EnvironmentError("%s not found" % (_mysql_config_path,))
    OSError: mysql_config not found
    ----------------------------------------
[0m[91mERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
[0m[91mTraceback (most recent call last):
  File "/home/odoo/maintainer-quality-tools/travis/clone_oca_dependencies", line 125, in <module>
    run(deps_checkout_dir, build_dir)
  File "/home/odoo/maintainer-quality-tools/travis/clone_oca_dependencies", line 110, in run
    subprocess.check_call(command)
  File "/usr/lib/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['pip', 'install', '--no-binary', 'pycparser', '-Ur', '/home/odoo/dependencies/server-backend/requirements.txt']' returned non-zero exit status 1
```


You can try installing `libmysqlclient-dev` apt package like original project:
 - https://github.com/OCA/server-backend/blob/c1349bb4bcab78113032f0bd867077b4269f43e1/.travis.yml#L15

**Comment by pedrobaeza** [comment]
Thank you very much for the info, @moylop260

**Comment by chienandalu** [comment]
Thanks @moylop260 !

File: .travis.yml [hunk 1]
```diff
@@ -16,6 +16,7 @@ addons:
 #      https://github.com/travis-ci/apt-source-whitelist/blob/master/ubuntu.json
     packages:
       - expect-dev  # provides unbuffer utility
+      - libmysqlclient-dev # To install mysqlclient and import MySQLdb
```
Comment (nhomar) [inline .travis.yml:19]: Why?

Comment (pedrobaeza) [reply .travis.yml:19]: See @moylop260's comment: https://github.com/OCA/account-invoicing/pull/586#issuecomment-524364832

**Comment by nhomar** [comment]
Sorry for the ackward question.

But this module looks a little over-designed.

1. why an extra table to set a simple discount amount?
2. All the computation is overwritten, when this can be done with an extra account move line (representing the global discount).
3. The new model is to restrict amounts (I got that AFAICU) but why not simple do it by amount in the the invoice and then raise a message to avoid or an onchange to pre-compute.

They are honest questions, but too much complicated on my opinion.

**Comment by pedrobaeza** [comment]
Hi @nhomar don't hesitate to ask of course. I'll try to answer:

> 1.  why an extra table to set a simple discount amount?

There can be more than one global discount, and they are stacked. Imagine the difficulty of computing this by hand, with hundred of lines and several stacked discounts.

> 2. All the computation is overwritten, when this can be done with an extra account move line (representing the global discount).

We needed to hack a bit `account` module compute methods for achieving the desired result, but it's done in the most compatible way ala [Tecnative style](https://www.youtube.com/watch?v=Hq9IbN7F9cg) :wink:

An extra invoice line doesn't serve because we need to affect the base amount for all the taxes in the invoice. Imagine that you have an invoice with 3 different taxes percentages. Doing "by hand" will require at least the same 3 number of discount lines, and VAT reports will be very dirty due to the extra discount lines.

> 3. The new model is to restrict amounts (I got that AFAICU) but why not simple do it by amount in the the invoice and then raise a message to avoid or an onchange to pre-compute.

I don't get you in this point, but I hope that previous explanations explain this also.

**Comment by rafaelbn** [comment]
@chienandalu David please check true the dependency in the descriptión

**Review by rafaelbn — APPROVED** [review]
Tested functionaly :+1:

**Comment by rafaelbn** [comment]
@carlosdauden @sergio-teruel please could you check this PR please?

File: account_global_discount/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2019 Tecnativa S.L. - David Vidal
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    'name': 'Base Global Discount',
+    'version': '11.0.1.0.0',
+    'category': 'Point of Sale',
```
Comment (sergio-teruel) [inline account_global_discount/__manifest__.py]: ¿Point of sale?

Comment (chienandalu) [reply account_global_discount/__manifest__.py]: Sure :sweat_smile:

**Comment by pedrobaeza** [comment]
What is missing here?

**Comment by rafaelbn** [comment]
Then can we merge this one ? :+1: is in production several months and works like a charm!

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Yes merging as there's no more answers from @nhomar 

/ocabot merge

**Pull request merged**